### ISSUE TITLE
parallel distributed grid and MeshSmoothing

### ISSUE THREAD

**Issue by kronbichler** [dealii/dealii#1655]
When working with the following simple code:
```
#include <deal.II/base/mpi.h>
#include <deal.II/grid/grid_generator.h>
#include <deal.II/distributed/tria.h>

int main(int argc, char** argv)
{
  dealii::Utilities::MPI::MPI_InitFinalize mpi(argc, argv, 1);

  dealii::parallel::distributed::Triangulation<2>
    tria(MPI_COMM_WORLD, dealii::Triangulation<2>::maximum_smoothing);
  dealii::GridGenerator::hyper_cube(tria,0,1);

  tria.refine_global(4);
}
```
I get the following error when running on two processors:
```
An error occurred in line <12546> of file <source/grid/tria.cc> in function
    bool dealii::Triangulation<2, 2>::prepare_coarsening_and_refinement()
The violated condition was: 
    cell_is_patch_level_1(cell)
```
The problem appears to be that the check for patch_level one that assumes that non-active cells either have no children active or all of the their children active fails for distributed triangulations where some cells are artificial. I first thought that simply modifying the checks for is_patch_level_1 to ignore artificial cells would be enough but then there are all sorts of accidents happening. Since I know too little about the workings of the grid (and in particular why the parallel distributed code actually ends up in this section here since it is up to p4est to decide about the grid), I hope for someone with more knowledge to comment on it. Shouldn't we ignore the maximum smoothing somehow? Why has nobody else seen such problems?

**Comment by tjhei** [comment]
> Shouldn't we ignore the maximum smoothing somehow?

Yes, this is clearly broken and we should assert or ignore some of the smoothing flags.

> and in particular why the parallel distributed code actually ends up in this section here since it is up to p4est to decide about the grid

p4est has fewer options regarding smoothing than we support for our Triangulation. If we make our local flags for refine/coarsen conform to some smoothing flags before sending them to p4est we might (but not in all cases) get a mesh that conforms to them globally. To be honest: I think none of the smoothing flags work as expected right now.

> Why has nobody else seen such problems?

smoothing flags haven't been tested at all... :-(

**Comment by bangerth** [comment]
Since we can't enforce any smoothing across processor boundaries, I would not be surprised if there are places in `prepare_c_and_r()` that *assume* that a smoothing property is true (because it was enforced the last time the mesh was refined) but where that property is not actually satisfied. In the best case, this would lead to a violated internal assertion, and in the worst case simply wrong code.

I dimly remember that for exactly these reasons, we did not allow setting smoothing flags at all when `p::d::Triangulation` was written originally. Does anyone remember when and why that was changed?

**Comment by tjhei** [comment]
I think we always allowed these flags and we are at least using ```smoothing_on_refinement | smoothing_on_coarsening``` in ASPECT.

**Comment by tjhei** [comment]
We will support ```limit_level_difference_at_vertices``` after #1651 is merged (so far we always acted as if this was set), we should support all other flags (with the caveat that it might be violated globally) except ```patch_level_1``` and ```coarsest_level_1```. My vote would be to simply disable those two (otherwise maximum_smoothing won't work).

**Comment by kronbichler** [comment]
For me, this sounds as a reasonable approach.

**Comment by bangerth** [comment]
#1651 was merged -- @kronbichler, does this resolve at least part of the issue? I guess we probably still need to add assertions about all of the other smoothing flags we don't currently support, right?

**Comment by tjhei** [comment]
huh? I think #1651, #1676, and #1711 are still open.

**Comment by bangerth** [comment]
Oh, I'm mistaken. #1651 was simply closed, not merged.

**Comment by gfcas** [comment]
Dear all,
we just encountered this problem in our application code, for a quasi identical situation as described above. Is there a current workarround or solution for this issue available? For the moment we excluded ``patch_level_1`` and ``coarsest_level_1`` flags and this worked for us.
